processor/k8sattributesprocessor/internal/kube/client.go

```diff
@@ -183,110 +204,134 @@ func New(
-				originalReplicaset, success := object.(*apps_v1.ReplicaSet)
-				if !success { // means this is a cache.DeletedFinalStateUnknown, in which case we do nothing
-					return object, nil
-				}
-
-				return removeUnnecessaryReplicaSetData(originalReplicaset), nil
-			},
-		)
-		if err != nil {
-			return nil, err
-		}
 	}
 
 	if c.extractNodeLabelsAnnotations() || c.extractNodeUID() {
-		c.nodeInformer = k8sconfig.NewNodeSharedInformer(c.kc, c.Filters.Node, 5*time.Minute)
+		if informersFactory.newNodeInformer == nil {
+			informersFactory.newNodeInformer = newNodeSharedInformer
+		}
 	}
 
 	if c.extractDeploymentLabelsAnnotations() {
-		c.deploymentInformer = newDeploymentSharedInformer(c.kc, c.Filters.Namespace)
+		if informersFactory.newDeploymentInformer == nil {
+			informersFactory.newDeploymentInformer = newDeploymentSharedInformer
+		}
 	}
 
+	c.informerFactories = informersFactory
+
 	return c, err
 }
 
 // Start registers pod event handlers and starts watching the kubernetes cluster for pod changes.
 func (c *WatchClient) Start() error {
+	var err error
+	c.informerMut.Lock()
+	defer c.informerMut.Unlock()
+
+	// if we return an error, we need to signal any informers we created to stop
+	defer func() {
+		if err != nil {
+			close(c.stopCh)
+		}
+	}()
+
+	// Set up handlers for other informers before the Pod informer, as Pod metadata depends on them.
 	synced := make([]cache.InformerSynced, 0)
-	// start the replicaSet informer first, as the replica sets need to be
-	// present at the time the pods are handled, to correctly establish the connection between pods and deployments
-	if c.Rules.DeploymentName || c.Rules.DeploymentUID {
-		reg, err := c.replicasetInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+
+	// set up the replicaset informer if needed
+	if c.informerFactories.newReplicaSetInformer != nil {
+		transformFunc := func(object any) (any, error) {
+			originalReplicaset, success := object.(*apps_v1.ReplicaSet)
+			if !success { // means this is a cache.DeletedFinalStateUnknown, in which case we do nothing
+				return object, nil
+			}
+
+			return removeUnnecessaryReplicaSetData(originalReplicaset), nil
+		}
+		c.replicasetInformer, err = c.informerFactories.newReplicaSetInformer(c.kc, c.Filters.Namespace, transformFunc, c.stopCh)
+		if err != nil {
+			return err
+		}
+		c.replicasetHandlerRegistration, err = c.replicasetInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
 			AddFunc:    c.handleReplicaSetAdd,
 			UpdateFunc: c.handleReplicaSetUpdate,
 			DeleteFunc: c.handleReplicaSetDelete,
 		})
 		if err != nil {
 			return err
 		}
-		synced = append(synced, reg.HasSynced)
-		go c.replicasetInformer.Run(c.stopCh)
+		synced = append(synced, c.replicasetHandlerRegistration.HasSynced)
 	}
 
-	reg, err := c.namespaceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+	// set up the namespace informer
+	c.namespaceInformer = c.informerFactories.newNamespaceInformer(c.kc, c.stopCh)
```

Comment:
Based on https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/36234#issuecomment-2929767810, is the plan to eventually do the initialisation of `namespaceInformer` (and others) in the component's factory?

Reply:
Logically, that's what happens here as well, but the factory is just a function that creates and starts the informer naively. When I introduce the shared factory in a follow-up, this code won't change, we'll just pass different functions in `informerFactories` which will internally use the shared factory object.